lib/src/main/java/io/ably/lib/transport/ConnectionManager.java

```diff
@@ -301,6 +302,7 @@ StateIndication validateTransition(StateIndication target) {
 
         @Override
         StateIndication onTimeout() {
+            this.timeout = ReconnectionStrategy.getRetryTime(ably.options.disconnectedRetryTimeout, ++disconnectedRetryAttempt);
```

Comment:
I'm not sure if moving this here is going to work as expected? If I'm reading this correctly:

- We transition to disconnected
- We have some number of timeouts (where the state goes disconnected -> connecting -> disconnected etc etc)
- We connect successfully
- We disconnect again... and now the timeout value is the same as it was at the last timeout in `disconnected`, and won't go back to the "correct" value until the first timeout happens again.

Whereas before - this would be handled and the first transition from disconnected to something else and then back again would set the retry delay back to its "first" value.

Reply:
When we connect again, `disconnectedRetryAttempt` will be set to zero, so it will re-generate the timeout value again.

Reply:
I think I got your point, till first timeout it will use the old value from last timeout. After that it will start following the right sequence.
But, I don't agree that it used to work before. We had a clear bug that was not able to set timeout value at right place.

Reply:
I will reproduce this scenario using tests.

Reply:
That's correct - but we're not resetting the `timeout` value on the disconnected state when we transition back to `connected`.

The initial value set is in the constructor - and we only call that constructor once, in the ConnectionManager constructor.

The only other time it's set is when the Disconnected state times out (and we calculate the new value based on the number of attempts). For the first time we move back to disconnected again after being connected, the timeout value will be whatever it was the last time we were disconnected?

Reply:
I have done the fix and pushing it soon.

Reply:
Fixed.

Reply:
I will check the scenario where there is forced `instant retry`. It should atleast not increment timeout value at that case

Reply:
Will this work if there's a delay in the state being enacted in a SynchronousStateChangeAction from onTransportUnavailable?

In this case:

- setState would be called in the SynchronousStateChangeAction constructor, with the `timeout` still set to whatever it was previously, which would set `currentState`
- The state change action would then be processed by the action loop (and still waiting on the `currentState` timeout)

Reply:
Currently, the tests are passing, I will try testing this scenario too. 
@AndyTWF thanks for thorough review of the PR

Reply:
Okay, fixed!